src/Core/Types/DataType.cs

```diff
@@ -89,7 +89,19 @@ public DataType Clone()
             return dt as T;
         }
 
-		protected void ThrowBadSize()
+        public T TypeReferenceAs<T>() where T : DataType
+        {
+            DataType dt = this;
+            TypeReference typeRef = dt as TypeReference;
+            while (typeRef != null)
+            {
+                dt = typeRef.Referent;
+                typeRef = dt as TypeReference;
+            }
+            return dt as T;
+        }
+
```

Comment:
This looks very similar to `DataType.ResolveAs`. Was there a particular reason why you couldn't use it? It seems to me that `DataType.ResolveAs` could be improved to handle this case as well.